optaplanner-examples/src/main/java/org/optaplanner/examples/flightcrewscheduling/persistence/FlightCrewSchedulingGenerator.java

```diff
@@ -0,0 +1,298 @@
+        List<Flight> flightList = new ArrayList<>(flightListSize);
+        List<FlightAssignment> flightAssignmentList = new ArrayList<>(flightListSize * EMPLOYEE_COUNT_PER_FLIGHT);
+
+        List<Airport> airportList = solution.getAirportList();
+        Airport centerAirport = airportList.get(0);
+        int homeAirportListSize = Math.min(airportList.size() / 10, flightRoundTripsPerDay / 5);
+        homeAirportList = airportList.stream()
+                .sorted(Comparator.comparingDouble(centerAirport::getHaversineDistanceInKmTo))
+                .limit(homeAirportListSize)
+                .collect(Collectors.toList());
+        LocalDate firstDate = solution.getScheduleFirstUTCDate();
+        LocalDate lastDate = solution.getScheduleLastUTCDate();
+        int flightNumberSuffix = 1;
+        long flightId = 0L;
+        long flightAssignmentId = 0L;
+        for (int i = 0; i < flightRoundTripsPerDay; i++) {
+            int flightCount = i % 2 == 0 ? 2 : 3;
+            ArrayList<Airport> selectedAirports = new ArrayList<>(flightCount);
+            Airport firstAirport = extractRandomElement(random, homeAirportList);
+            selectedAirports.add(firstAirport);
+            ArrayList<Airport> nonFirstAirports = new ArrayList<>(airportList);
+            nonFirstAirports.remove(firstAirport);
+            Collections.shuffle(nonFirstAirports, random);
+            selectedAirports.addAll(nonFirstAirports.subList(0, flightCount - 1));
+            for (int j = 0; j < flightCount; j++) {
+                String flightNumber = "AB" + String.format("%03d", flightNumberSuffix);
+                flightNumberSuffix++;
+                Airport departureAirport = selectedAirports.get(j);
+                Airport arrivalAirport = selectedAirports.get((j + 1) % flightCount);
+                int flyingTime = (int) ((departureAirport.getHaversineDistanceInKmTo(arrivalAirport)
+                        / PLANE_SPEED_IN_KM_PER_MINUTE) + PLANE_TAKE_OFF_AND_LANDING_MINUTES);
+                int departureMinute = START_MINUTE_OF_DAY + random.nextInt(
+                        END_MINUTE_OF_DAY - flyingTime - START_MINUTE_OF_DAY + 1);
+                int arrivalMinute = departureMinute + flyingTime;
+                for (LocalDate date = firstDate; date.compareTo(lastDate) <= 0; date = date.plusDays(1)) {
+                    Flight flight = new Flight();
+                    flight.setId((long) flightId);
+                    flightId++;
+                    flight.setFlightNumber(flightNumber);
+                    flight.setDepartureAirport(departureAirport);
+                    flight.setDepartureUTCDateTime(date.atTime(departureMinute / 60, departureMinute % 60));
+                    flight.setArrivalAirport(arrivalAirport);
+                    flight.setArrivalUTCDateTime(date.atTime(arrivalMinute / 60, arrivalMinute % 60));
+                    for (int indexInFlight = 0; indexInFlight < EMPLOYEE_COUNT_PER_FLIGHT; indexInFlight++) {
```

Comment:
4th level of nested for loop suggests there is a space for refactoring.

Reply:
You have a good point,
but due to the flightAssignmentId and the flightId (which need to be returned from methods) this impacts readability too.
In general, I am more lenient on method length and nesting dept for Generator and FileIO code than I am for other code, because it's really just handling each property one by one.

Let me see if I can split off the FlightAssignment creation entirely, instead of just extracting method of that 4th for loop.

Reply:
Yes, that worked like a charm, good suggestion!